src/clarity/debug/dap/mod.rs

```diff
@@ -0,0 +1,1146 @@
+            variables.push(Variable {
+                name: name.to_string(),
+                value: value.to_string(),
+                var_type: Some(type_for_value(value)),
+                presentation_hint: None,
+                evaluate_name: None,
+                variables_reference: 0,
+                named_variables: None,
+                indexed_variables: None,
+                memory_reference: None,
+            });
+        }
+
+        // Variables
+        for (name, metadata) in &contract_context.meta_data_var {
+            let data_types = contract_context.meta_data_var.get(name).unwrap();
+            let value = global_context
+                .database
+                .lookup_variable(
+                    &contract_context.contract_identifier,
+                    name.as_str(),
+                    data_types,
+                )
+                .unwrap();
+            variables.push(Variable {
+                name: name.to_string(),
+                value: value.to_string(),
+                var_type: Some(format!("{}", metadata.value_type)),
+                presentation_hint: None,
+                evaluate_name: None,
+                variables_reference: 0,
+                named_variables: None,
+                indexed_variables: None,
+                memory_reference: None,
+            });
+        }
+
+        // Maps
+        for (name, metadata) in &contract_context.meta_data_map {
+            // We do not grab any values for maps. Users can query map values in the console.
+            let map_type = format!("{{{}: {}}}", metadata.key_type, metadata.value_type);
+            variables.push(Variable {
+                name: name.to_string(),
+                value: map_type.clone(),
+                var_type: Some(map_type),
+                presentation_hint: None,
+                evaluate_name: None,
+                variables_reference: 0,
+                named_variables: None,
+                indexed_variables: None,
+                memory_reference: None,
+            });
+        }
+        self.variables.insert(scope_id, variables);
+
+        self.scopes.insert(stack_frame.id, scopes);
+    }
+}
+
+impl EvalHook for DAPDebugger {
+    fn begin_eval(
+        &mut self,
+        env: &mut Environment,
+        context: &LocalContext,
+        expr: &SymbolicExpression,
+    ) {
+        let source = Source {
+            name: Some(env.contract_context.contract_identifier.to_string()),
+            path: Some(
+                match self
+                    .contract_id_to_path
+                    .get(&env.contract_context.contract_identifier)
+                {
+                    Some(path) => path.clone(),
+                    _ => "debugger".to_string(),
+                },
+            ),
+            source_reference: None,
+            presentation_hint: None,
+            origin: None,
+            sources: None,
+            adapter_data: None,
+            checksums: None,
```

Comment:
Speaking of checksums, I was actually wondering about this, how is the DAP protocol dealing with "stale" data? Let say I add a breakpoint using VSCode UI, delete the line, are we updating the state automatically?

Reply:
Good question. No, currently, if a user changes the source in the middle of a debug session, the original code is still debugged. That would be a cool feature to support later. Opening another issue to remember that 😄